doc/rdebug.1
.\" $Id$
.TH rdebug 1 
.SH NAME
rdebug \- Fast Ruby debugger
.SH SYNOPSIS
.B rdebug 
[debugger-options] 
rdebug
[script-options...]
.SH "DESCRIPTION"
This manual page documents briefly the
.BR rdebug
command.
.PP
.B rdebug
is a fast implementation of the standard Ruby debugger debug.rb.  It
is implemented by utilizing a Ruby C API hook, allows for remote
debugging and can be used as the Ruby debugger backend interface for a
development environment.
.PP
The commands generally follow gdb's command set unless there's good
reason not to.

.PP
rdebug can do four main kinds of things (plus other things in support of
these) to help you catch bugs in the act:

.TP
\ \ \ \(bu
Start or restart your Ruby script, specifying arguments that might
affect its behavior.

.TP
\ \ \ \(bu
Make your program stop at various points possibly determined by
specified conditions.

.TP
\ \ \ \(bu
Examine what has happened when your program has stopped.

.TP
\ \ \ \(bu
Change things in your program, so you can experiment with correcting the
effects of one bug and go on to learn about another.
.PP

Here are some of the most frequently-needed commands:
.TP
.B break \fR[\|\fIfile\fB:\fIline\fR\fR|\fIclass.method\fR] \fR[if \fIexpr\fR]
\&
Set a breakpoint at \c
.I class.method\c
\& or at the specified file and line.
.TP
.B continue \fR[\fIline\fR]
Continue running your program (after stopping, e.g. at a
breakpoint). If a line is given a one-time breakpoint is set there.
.TP
.B delete \fR[\fIbreakpoint-numbers\fR]
\&
Delete breakpoints by number. If no number is given delete all breakpoints.
.TP
.B down \fR[\|\fIcount\fR\|]
Move down one block frame. If count is given move up that many frames. A negative number
goes the other direction and is like the up command
.TP
.B finish
Run until the completion of the current function or method.
.TP
.BI frame " frame-number"
Set the stack frame to \fIframe-number\fR for purposes of examinine local variables. For positioning relative to the current frame, use 
.B up
or 
.B down. A negative number starts counting from the other end.
.TP
.B help \fR[\|\fIname\fR\|]
Show information about rdebug command \c
.I name\c
\&, or general information
about using rdebug.
.TP
.B info \fR[\|\fIname\fR\|]
Get the various information usually about the debugged program.
.TP
.B irb \fIcommand\fR
Run an interactive ruby shell (irb) using the current environment.
.TP
.B list \fR[\|\fIfile\fB:\fIline\fR|\fIfunction]
type the text of the program in the vicinity of where it is presently stopped
or at the specified function or file and line.
.TP
.B next \fR[\|\fIcount\fR\|]
Execute next program line(s) (after stopping); step \c
.I over\c
\& any
function calls in the line.
.TP
.BI pp " expr"\c
\&
Pretty print the value of an expression.
.TP
.BI print " expr"\c
\&
Display the value of an expression.
.TP
.BI ps " expr"\c
\&
Print an array as a columized sorted list.
.TP
.B quit
Exit from the debugger.
.TP
.B run \fR[\|\fIarglist\fR\|]
(Re)start your program (with \c
.I arglist\c
\&, if specified). If you want the debugger to get reloaded, use
.B restart
instead.
.TP
.B set
Modify parts of the debugger environment.
.TP
.B show
See the debugger environment settings
.TP
.BI source " filename"\c
\&
Read and execute the lines in file \fIfilename\fR as a series of debugger 
commands.
.TP
.B step \fR[\|\fIcount\fR\|]
Execute next program line(s) (after stopping); step \c
.I into\c
\& any
function calls in the line.
.TP
.B up \fR[\|\fIcount\fR\|]
Move up one block frame. If count is given move up that many frames. A negative number
goes the other direction and is like the down command
.TP
.B where \fR[\|\fIcount\fR\|]
Display all or \fIcount\fR items of the program stack.
.PP
For full details on rdebug, see \c
http://rubyforge.org/projects/ruby-debug/
.SH OPTIONS
.PP
.TP 10
.TP
.B \-A | \-\-annotate LEVEL
Set gdb-style annotation to LEVEL, a number. Additional information is output
automatically when program state is changed. This can be used by
front-ends such as GNU Emacs to post this updated information without
having to poll for it.
.TP
.B \-\-client
Connect to a remote debugger. Used with another rdebug invocation using \-\-server.
See also \-\-host and \-\-cport options
.TP
.B \-\-cport=PORT
Port used for control commands.
.TP
.B \-d | \-\-debug
Set $DEBUG true.
.TP
.B \-\-emacs
Activates GNU Emacs mode. Debugger prompts are prefaced with two octal
032 characters.
.TP
.B \-h | \-\-host=HOST
Host name used for remote debugging.
.TP
.B \-I | \-\-include PATH
Add PATH to $LOAD_PATH
.TP
.B \-m | \-\-post-mortem
Activate post-mortem mode.
.TP
.B \-\-no-control
Do not automatically start control thread.
.TP
.B \-\-no\-stop
Do not stop when script is loaded.
.TP
.B \-p | \-\-port=PORT
Host name used for remote debugging.
.TP
.B \-r | \-\-require SCRIPT
Require the library, before executing your script.
.TP
.B \-\-script FILE
Name of the script file to run.
.TP
.B \-x | \-\-trace
Show lines before executing them.
.TP
.B \-\-no\-quit
Do not quit when script terminates. Instead rerun the program.
.TP
.B \-\-version
Show the version number and exit.
.TP
.B \-\-verbose
Turn on verbose mode.
.TP
.B \-\-v
Print the version number, then turn on verbose mode if a script name
is given. If no script name is given just exit after printing the
version number.
.TP
.B \-\-nx
Don't execute commands found in any initialization files, e.g. .rdebugrc.
.TP
.B \-\-keep-frame-binding
Keep frame bindings.
.TP
.B \-\-script=FILE
Name of the script file to run
.B \-s | \-\-server
Listen for remote connections. Another rdebug session accesses using the \-\-client option.
See also the \-\-host, \-\-port and
\-\-cport options
.TP
.B \-w | \-\-wait
Wait for a client connection, implies -s option.
.TP
.B \-\-help
Show invocation help and exit.
.PD
.SH "SEE ALSO"
.Sp
http://rubyforge.org/projects/ruby-debug/
.SH AUTHOR
rdebug was written by Kent Siblev. This manual page was written by
Rocky Bernstein <rocky@gnu.org>


CHANGES
0.10.1
- Experimental: 
  remove Debugger.stop on exit exception raise. May allow catchpoints
  to work and tracing user code handling Exit exceptions
- add "toggle" command
- fix bug in "info variables" when string had embedded %s'
- Number keys move stack position in Emacs stack secondary buffer.
- --emacs is now --emacs-basic while --emacs now implies
  --emacs-basic --annotate=3 --post-mortem --no-control --no-start --no-quit
- split off breakpoint code in ruby_debug.c; preface ruby_debug global 
  Ruby variables with rdebug_.
0.10.0
- '-r' option can be used to require additional libraries.
- --noquit option added to stay in debugger when the program exits
- gdb-like --annotate option added. Can be used by front-ends to get information
  without polling
- Fixed 'var const' command. Issue #10847.
- Using pretty-print for all var commands.
- Better error reporting for commands that require a numeric argument.
- Fixed Kernel#binding_n method
- Add option -d ($DEBUG = true) and --verbose. Make -v work like ruby.
- Remove debugger messages caused when warnings are turned on.
- "info" and "show" commands added. "set" made more like gdb's
  set. subcommands can be abbreviated and are case insensitive.
- restart program if it terminates normally and we've got a tty and 
  we stop on the first statement. 
- help is in tidy column format. method lists are shown that way as well.
  the "width" setting ("set/show width") is used  for the line width
- stack traces now show parameter names and types. "info args" lists just
  the parameters (with the most recent values, not the values at call time).
- post-mortem "exit" bug fixed. 
- More Emacs-friendly: rdebug-track.el will track location inside an Emacs
  shell. Emacs position information is shown in breakpoints and catchpoints
  similar to gdba. Commands to position in another window a unit test traceback
  or ruby traceback. Much more work invisioned for Emacs.
- INCOMPATIBLE CHANGE: "break" now sets a breakpoint on the current line
  (same as gdb). Use "info break" for a list of breakpoints.
- INCOMPATIBLE CHANGE: "script" command removed. Use "source" command instead
  (same as gdb).
- Run .rdebugrc on Debugger.start. Look for a file in the current directory and
  run that instead the one in $HOME if that exists. Again, inspired and compatible
  with gdb.
- Changes compatible with Ruby 1.9. NOTE: this debugger will NOT work with
  Ruby 1.9
- leaving irb shows position same as entering debugger; "list" position
  is also cleared when leaving irb
- help "foo" gives message "Undefined command "foo" rather than a list
  of help commands. (Message test is gdb's)
- Add set linetrace+ - similar step+ for linetrace
- Start unit tests.
- Start a reference guide.

0.9.3
- Fixed if..elsif..end stepping.
- From irb session Ctrl-C or 'cont' command continues execution without showing the debugger prompt.
- Added Debugger.settings method to programatically modify command settings.
- Added Kernel#breakpoint as alias to Kernel#debugger is the former is not already defined.

0.9.2
- Fixed file comparison in Windows platform.
- Added setter methods to Breakpoint properties
- Added breakpoint hit condition functionality (not available via CLI yet) and methods:
    Breakpoint:hit_count
    Breakpoint:hit_value[=]
    Breakpoint:hit_condition[=]

0.9.1
- Fixed incorrect stack calculation.
- Context#stop_next= method aliased as Context#step.
- Added the 'force' parameter to Context#step_over.
- Added the 'force' parameter to Context#step.
- 'next+/step+' commands forces to move to another line
- Added a new 'forcestep' setting.

0.9
- Kernel#debugger method will start the debugger if it's not running.
- Added Context#stop_reason method.
- Calling a method with a block will create a new frame. This changes the behavior of 'next' command. So in order to step into a block, 'step' command must be used. That fixes bug #9629.
- Added the possibility to add a temporary context-specific breakpoint. Context#breakpoint and Context#set_breakpoint methods are added. 
- 'cont' command now accepts a numerical parameter which implements 'Continue until line' behavior.
- Added new Context.frame_class method
- Added new 'framefullpath' setting.
- Added new 'frameclassname' setting.
- All Ruby's 'eval' and require/load methods create a new frame. Fixes bug #9686.

0.8.1
- Added a shortcut module 'debugger'. require "ruby-debug/debugger" will start the debugger and stop at the next line (similar to require 'debug').
- Fixed remote debugging.

0.8
- Extract the base debugger API into a separate gem (ruby-debug-base), so it will be easier to add a new interface.
- Added 'set autoirb' setting.
- Bugfixes.

0.7.5
- Fixed 'reload on' command
- 'reload on' command is removed in favor of 'set autoreload'
- rdebug will evaluate ~/.rdebugrc script on startup

0.7.4
- Added a workaround of the Ruby interpreter problem where a method created with Module#define_method 
  and which raises an exception doesn't trigger a :return event, this way screwing the stack trace.
- Fixed a situation of an array 'out of bounds' access.
- Fixed the help for 'where' command.

0.7.3
- Fixed a case when a frame is not popped up properly.
- Removed Context.ignore= method, since it can result with the segmentation fault error.
- Fixed the case when Context#suspend may effect the state of the thread on Context#resume
- Fixed several cases of seg faults when accessing dyna_vars structure.

0.7.2
- Fixed Context#resume (a thread should be waked up only when it was running when it was suspended).
- When handling post-mortem exception, all threads must be suspended.

0.7.1
- Fixed 'delete' command

0.7
- Eliminated explicit Frame object. Use Context.frame_[binding,file,line] instead.
- Fixed help command.
- Renamed Debugger.keep_frame_info to Debugger.keep_frame_binding
- 'eval' command is available, even when keep_frame_binding is not used.
- New 'set' command is available.

0.6.2
- Added thread lookup cache.

emacs/rdebug.el
;;; rdebug.el --- Ruby debugger user interface, startup file.

;; Copyright (C) 2006, 2007, 2008 Rocky Bernstein (rocky@gnu.org)
;; Copyright (C) 2007, 2008 Anders Lindgren

;; $Id: rdebug.el 409 2007-12-14 02:36:37Z rockyb $

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;; Boston, MA 02111-1307, USA.

;;; Commentary:

;; This file contains the initialization needed for the Emacs user
;; interface to the `rdebug' Ruby debugger.
;;
;; To use this package, place the following line in an appropriate
;; init file (for example ~/.emacs):
;;
;;    (require 'rdebug)
;;

;;; Code:


;; -------------------------------------------------------------------
;; Consistency checks.
;;

(if (< emacs-major-version 22)
    (error
     "This version of rdebug.el needs at least Emacs 22 or greater - you have version %d."
     emacs-major-version))


;; -------------------------------------------------------------------
;; Support functions.
;;

(defun rdebug-directory ()
  "The directory of this file, or nil."
  (let ((file-name (or load-file-name
                       (symbol-file 'rdebug-directory))))
    (if file-name
        (file-name-directory file-name)
      nil)))


(defun rdebug-compare-filenames (f1 f2)
  "Canonicalize and compare file names."
  ;; Canonicalize by:
  ;;  1) file-truename ensures that the file has got the correct case,
  ;;     and that "..":s in the path are eliminated.
  ;;  2) file-name-as-directory ensures "/foo" and "/foo/" becomes equal.

  ;; Note: for some reason, when the `comp-elisp' external program is
  ;; used, `nil' is part of `load-path'.
  (if f1
      (setq f1 (file-name-as-directory (file-truename f1))))
  (if f2
      (setq f2 (file-name-as-directory (file-truename f2))))
  (equal f1 f2))


;; Add the directory of `rdebug.el' to the load-path. This ensures
;; that all the user have do to use this package is to load this file.
(let ((dir (rdebug-directory)))
  (if dir
      (add-to-list 'load-path dir nil 'rdebug-compare-filenames)))


;; -------------------------------------------------------------------
;; Autoloads.
;;

(autoload 'rdebug "rdebug-core"
  "Run the rdebug Ruby debugger and start the Emacs user interface.

By default, the \"standard\" user window layout looks like the following:

+----------------------------------------------------------------------+
|                                Toolbar                               |
+-----------------------------------+----------------------------------+
| Debugger shell                    | Variables buffer                 |
+-----------------------------------+----------------------------------+
|                                   |                                  |
| Source buffer                     | Output buffer                    |
|                                   |                                  |
+-----------------------------------+----------------------------------+
| Stack buffer                      | Breakpoints buffer               |
+-----------------------------------+----------------------------------+

The variable `rdebug-many-windows-layout-function' can be
customized so that another layout is used. In addition to a
number of predefined layouts it's possible to define a function
to perform a custom layout.

If `rdebug-many-windows' is nil, only a traditional debugger
shell and source window is opened.

The directory containing the debugged script becomes the initial
working directory and source-file directory for your debugger.

The custom variable `gud-rdebug-command-name' sets the command
and options used to invoke rdebug." t)


(autoload 'rdebug-turn-on-debugger-support "rdebug-source"
  "Enable extra source buffer support for the `rdebug' Ruby debugger.

This includes a 'Debugger' menu and special key bindings when the
debugger is active."
  t)


(autoload 'rdebug-track-attach "rdebug-track"
  "Do things to make the current process buffer work like a
rdebug command buffer." t)

(autoload 'turn-on-rdebug-track-mode "rdebug-track"
  "Turn on rdebugtrack mode.

This function is designed to be added to hooks, for example:
  (add-hook 'comint-mode-hook 'turn-on-rdebugtrack-mode)"
  t)


(add-hook 'ruby-mode-hook 'rdebug-turn-on-debugger-support)


;; -------------------------------------------------------------------
;; User definable variables
;;

(defcustom gud-rdebug-command-name
  "rdebug --emacs"
  "File name for executing the Ruby debugger and command options.
This should be an executable on your path, or an absolute file name."
  :type 'string
  :group 'gud)

(defcustom rdebug-line-width 120
  "Length of line before truncation occurs.
This value limits output in secondary buffers."
  :type 'integer
  :group 'rdebug)

(defcustom rdebug-many-windows t
  "*If non-nil, use the full debugger user interface, see `rdebug'.

However only set to the multi-window display if the rdebug
command invocation has an annotate options (\"--annotate 3\")."
  :type 'boolean
  :group 'rdebug)

(defcustom rdebug-populate-common-keys-function
  'rdebug-populate-common-keys-standard
  "The function to call to populate key bindings common to all rdebug windows.
This includes the secondary windows, the debugger shell, and all
Ruby source buffers when the debugger is active.

This variable can be bound to the following:

* nil -- Don't bind any keys.

* `rdebug-populate-common-keys-standard' -- Bind according to a widely used
  debugger convention:

\\{rdebug-example-map-standard}

* `rdebug-populate-common-keys-eclipse' -- Bind according to Eclipse.

\\{rdebug-example-map-eclipse}

* `rdebug-populate-common-keys-netbeans' -- Bind according to NetBeans.

\\{rdebug-example-map-netbeans}

* Any other value is expected to be a callable function that takes one
  argument, the keymap, and populates it with suitable keys."
  :type 'function
  :group 'rdebug)

(defcustom rdebug-restore-original-window-configuration :many
  "*Control if the original window layout is restored when the debugger exits.
The value can be t, nil, or :many.

A value of t means that the original layout is always restored,
nil means that it's never restored.

:many means that the original layout is restored only when
`rdebug-many-windows' is used."
  :type '(choice (const :tag "Always restore" t)
		 (const :tag "Never restore" nil)
		 (const :tag "Restore in many windows mode" :many))
  :group 'rdebug)

(defcustom rdebug-use-separate-io-buffer t
  "Non-nil means display output from the debugged program in a separate buffer."
  :type 'boolean
  :group 'gud)


(defcustom rdebug-window-layout-function
  'rdebug-window-layout-standard
  "*A function that performs the window layout of `rdebug'.

This is only used in `rdebug-many-windows' mode. This should be
bound to a function that performs the actual window layout. The
function should takes two arguments, the first is the source
buffer and the second the name of the script to debug.

Rdebug provides the following predefined layout functions:

* `rdebug-window-layout-standard'         -- See `rdebug'

* `rdebug-window-layout-conservative'     -- Source + Shell + Output

* `rdebug-window-layout-stack-of-windows' -- Extra windows to the right

* `rdebug-window-layout-rocky'            -- Rocky's own layout"
  :type
  '(choice
    (function :tag "Standard"         rdebug-window-layout-standard)
    (function :tag "Conservative"     rdebug-window-layout-conservative)
    (function :tag "Stack of windows" rdebug-window-layout-stack-of-windows)
    (function :tag "Rocky's own"      rdebug-window-layout-rocky)
    (function :tag "Other"            function))
  :group 'rdebug)


;; -------------------------------------------------------------------
;; The end.
;;

(provide 'rdebug)

;;; rdebug.el ends here
